### ISSUE TITLE
analyzer rule flutter_style_todos has false positives

### ISSUE THREAD

**Issue by ragingsquirrel3** [dart-lang/linter#3919]
This tracker is for issues related to:

* Analyzer

* Dart SDK Version: Dart SDK version: 3.0.0-21.0.dev (dev) (Mon Dec 12 06:46:26 2022 -0800) on "linux_x64" (no issue on stable)
* Whether you are using Windows, MacOSX, or Linux (if applicable): had issue on mac and linux

I was using analyzer w flutter_style_todos rule on dev version and saw it flag some code that wasn't an actual TODO comment, just a comment that happened to have the word TODO in it.

Comment looked this this:
```dart
  /// Generates a request for a single instance of the `modelType` by `id`.
  ///
  /// ```dart
  /// final request = ModelQueries.get(Todo.classType, 'some-todo-id-123');
  /// ```
```

I would not expect it to consider this a TODO comment.

**Comment by ragingsquirrel3** [comment]
Example of error if helpful https://github.com/aws-amplify/amplify-flutter/actions/runs/3708128949/jobs/6285349082#step:9:10

**Comment by pq** [comment]
Really helpful, thanks a million @ragingsquirrel3!

**Comment by AlexV525** [comment]
Although the comment looks weird in the editor, I cannot reproduce this on Dart 2.18.6. I suspect there is a regression with the analyzer, or please correct me if my code is invalid @ragingsquirrel3. I'm going to test again on the latest SDKs.

![image](https://user-images.githubusercontent.com/15884415/208612591-2b989996-80a0-46d6-97f8-adc9e795ce43.png)

**Comment by AlexV525** [comment]
Weird. I cannot even reproduce this on Dart SDK version: 3.0.0-35.0.dev (dev) (Thu Dec 15 18:25:10 2022 -0800) on "windows_x64"

**Comment by ragingsquirrel3** [comment]
These were the actual lines that caused the error in flutter dev:

https://github.com/aws-amplify/amplify-flutter/blob/next/packages/api/amplify_api/lib/src/graphql/model_helpers/model_mutations.dart#L27

https://github.com/aws-amplify/amplify-flutter/blob/next/packages/api/amplify_api/lib/src/graphql/model_helpers/model_queries.dart#L52

Those are the lines in the build above though not sure what's special about them compared to line you tried.

**Comment by AlexV525** [comment]
> Those are the lines in the build above though not sure what's special about them compared to line you tried.

Thanks, I can reproduce using these lines.

**Comment by AlexV525** [comment]
The previous rule generates lints on lines like this as mentioned in https://github.com/dart-lang/linter/pull/1145#issuecomment-418918058:

```dart
/// Return all of the TODO comments in the file.
```

For me, I wouldn't treat it as a TODO comment. However, consider the comment in [diagnostics_test.dart#L567](https://github.com/flutter/flutter/blob/a9275d2154b16f2bb4ee6474f579d00f2d04c8d6/packages/flutter/test/foundation/diagnostics_test.dart#L567):

```dart
// There isn't anything interesting for this case as the children look the
// same with and without children. TODO(jacobr): this is an ugly test case. <-- Recognized as a TODO
// only difference is odd not clearly desirable density of B3 being right
// next to node C.
```

The above cases are all considered as TODOs. I'm not sure how should we deal at this moment. The PR fails for the same reason.

**Comment by srawlins** [comment]
We could probably quiet this down by not reporting when:

* 'TODO' is not at the beginning of a line of a comment, or
* 'TODO' is not at the beginning of a line of a comment, and is followed by whitespace?

**Comment by AlexV525** [comment]
> * 'TODO' is not at the beginning of a line of a comment, and is followed by whitespace?

Do you have an example for this?

**Comment by srawlins** [comment]
> > 'TODO' is not at the beginning of a line of a comment, and is followed by whitespace?

> Do you have an example for this?

The first example [above](https://github.com/dart-lang/linter/issues/3919#issuecomment-1361062855) is an example.

`/// Return all of the TODO comments in the file.`

**Comment by AlexV525** [comment]
https://dart-review.googlesource.com/c/sdk/+/363840

**Comment by bwilkerson** [comment]
@goderbauer

An opinion from the flutter team would be helpful. The proposal at this point is to not consider a comment to contain a TODO if there are any non-whitespace characters before the characters 'TODO'. Hence, a comment of the form
```
/// Return all of the TODO comments in the file.
```
would not be considered to be a TODO comment and would not be flagged by the lint.

Would such a change be acceptable?

**Comment by goderbauer** [comment]
I think that's acceptable.

**Comment by AlexV525** [comment]
cc @srawlins for reviewing the CL
